src/Data/String/NonEmpty.purs

```diff
@@ -0,0 +1,484 @@
+-- | Non-empty strings.
+-- |
+-- | Please note that the examples in this documentation use a notation like
+-- | `NonEmptyString "..."` for demonstration purposes, `NonEmptyString` cannot
+-- | be created directly like that, as we can't prove the string is non-empty to
+-- | the compiler at compile-time.
+module Data.String.NonEmpty
+  ( NonEmptyString
+  , NonEmptyReplacement(..)
+  , fromString
+  , fromCharArray
+  , singleton
+  , cons
+  , snoc
+  , fromFoldable1
+  , toString
+  , toCharArray
+  , charAt
+  , charCodeAt
+  , toChar
+  , appendString
+  , prependString
+  , contains
+  , indexOf
+  , indexOf'
+  , lastIndexOf
+  , lastIndexOf'
+  , uncons
+  , length
+  , localeCompare
+  , replace
+  , replaceAll
+  , take
+  , takeRight
+  , takeWhile
+  , drop
+  , dropRight
+  , dropWhile
+  , stripPrefix
+  , stripSuffix
+  , count
+  , splitAt
+  , toLower
+  , toUpper
+  , trim
+  , joinWith
+  , join1With
+  , joinWith1
+  , module Data.String
+  ) where
+
+import Prelude
+
+import Data.Foldable (class Foldable)
+import Data.Foldable as F
+import Data.Maybe (Maybe(..))
+import Data.Semigroup.Foldable (class Foldable1)
+import Data.Semigroup.Foldable as F1
+import Data.String (Pattern(..))
+import Data.String as String
+import Data.String.Unsafe as U
+import Unsafe.Coerce (unsafeCoerce)
+
+-- | A string that is known not to be empty.
+newtype NonEmptyString = NonEmptyString String
+
+derive newtype instance eqNonEmptyString ∷ Eq NonEmptyString
+derive newtype instance ordNonEmptyString ∷ Ord NonEmptyString
+derive newtype instance semigroupNonEmptyString ∷ Semigroup NonEmptyString
+
+instance showNonEmptyString :: Show NonEmptyString where
+  show (NonEmptyString s) = "(NonEmptyString.fromString " <> show s <> ")"
```

Comment:
`fromString` returns `Maybe NonEmptyString` so here I guess should be `unsafeFromString` or something like that.

Reply:
Hmm, yeah I guess so. I didn't want to add that originally as the `unsafeFromString` I had in mind was _really_ unsafe, but I guess if we do it as `fromJust <<< ...` and give it the `Partial` constraint that'll work well enough.